storage/database/sharded_database.go

```diff
@@ -151,102 +154,283 @@ func (pdb *shardedDB) Close() {
-	iterators []Iterator
-	key       []byte
-	value     []byte
+	shardedDBChanIterator
 
-	//numBatches uint
-	//
-	//taskCh   chan pdbBatchTask
-	//resultCh chan pdbBatchResult
+	resultCh chan entry
+	key      []byte // current key
+	value    []byte // current value
 }
 
 // NewIterator creates a binary-alphabetical iterator over the entire keyspace
 // contained within the key-value database.
-func (pdb *shardedDB) NewIterator() Iterator {
-	// TODO-Klaytn implement this later.
-	return nil
+// If you want to get unordered items faster in serial, checkout shardedDB.NewChanIteratorUnsorted().
+// If you want to get items in parallel from channels, checkout shardedDB.NewChanIterator()
+func (db *shardedDB) NewIterator() Iterator {
+	return db.newIterator(func(db Database) Iterator { return db.NewIterator() })
 }
 
 // NewIteratorWithStart creates a binary-alphabetical iterator over a subset of
 // database content starting at a particular initial key (or after, if it does
 // not exist).
-func (pdb *shardedDB) NewIteratorWithStart(start []byte) Iterator {
-	// TODO-Klaytn implement this later.
-	iterators := make([]Iterator, 0, pdb.numShards)
-	for i := 0; i < int(pdb.numShards); i++ {
-		iterators = append(iterators, pdb.shards[i].NewIteratorWithStart(start))
+func (db *shardedDB) NewIteratorWithStart(start []byte) Iterator {
+	return db.newIterator(func(db Database) Iterator { return db.NewIteratorWithStart(start) })
+}
+
+// NewIteratorWithPrefix creates a binary-alphabetical iterator over a subset
+// of database content with a particular key prefix.
+func (db *shardedDB) NewIteratorWithPrefix(prefix []byte) Iterator {
+	return db.newIterator(func(db Database) Iterator { return db.NewIteratorWithPrefix(prefix) })
+}
+
+func (db *shardedDB) newIterator(newIterator func(Database) Iterator) Iterator {
+	it := &shardedDBIterator{
+		db.NewChanIterator(context.Background(), newIterator),
+		make(chan entry, shardedDBCombineChanSize),
+		nil, nil}
+
+	go it.newCombineWorker()
+
+	return it
+}
+
+// newCombineWorker fetches any key/value from resultChs and put the data in resultCh
+// in binary-alphabetical order.
+func (it shardedDBIterator) newCombineWorker() {
+	entries := make([]entry, len(it.resultChs)) // contains smallest values from each iterators
+
+	// fill initial values for entries
+	for i := len(it.resultChs) - 1; i >= 0; i-- {
+		ch := it.resultChs[i]
+		e, ok := <-ch
+		if !ok { // channel is closed
+			it.resultChs = append(it.resultChs[:i], it.resultChs[i+1:]...)
+			entries = append(entries[:i], entries[i+1:]...)
+		} else {
+			entries[i] = e
+		}
 	}
 
-	for _, iter := range iterators {
-		if iter != nil {
-			if !iter.Next() {
-				iter = nil
+chanIter:
+	for len(it.resultChs) != 0 {
+		// check if done
+		select {
+		case <-it.ctx.Done():
+			logger.Trace("[shardedDBIterator] combine worker ends due to ctx")
+			break chanIter
+		default:
+		}
+
+		// look for smallest key
+		minEntry, minIdx := entries[0], 0
```

Comment:
How about using minheap for finding smallest key?

Reply:
Review reflected